src/tree/updater_quantile_hist.cc

```diff
@@ -88,14 +110,22 @@ void QuantileHistMaker::Update(HostDeviceVector<GradientPair> *gpair,
 bool QuantileHistMaker::UpdatePredictionCache(
     const DMatrix* data,
     HostDeviceVector<bst_float>* out_preds) {
-  if (!builder_ || param_.subsample < 1.0f) {
-    return false;
+  if (hist_maker_param_.single_precision_histogram) {
+    if (!float_builder_ || param_.subsample < 1.0f) {
+      return false;
+    } else {
+      return float_builder_->UpdatePredictionCache(data, out_preds);
+    }
   } else {
-    return builder_->UpdatePredictionCache(data, out_preds);
+    if (!double_builder_ || param_.subsample < 1.0f) {
+      return false;
+    } else {
+      return double_builder_->UpdatePredictionCache(data, out_preds);
+    }
```

Comment:
just extension for double and float builders

Reply:
Do we have a case when we have ```double_builder_``` and ```hist_maker_param_.single_precision_histogram == true ``` in the same time? I don't think

Reply:
I agree we don't, and this is not checked as you can see.
We still have to check `!float_builder_` as `QuantileHistMaker::UpdatePredictionCache` potentially can be called before first call `QuantileHistMaker::Update`. As I understand this check is about described check. (Similar applicable for double: if single_precision_histogram == true we check float_builder otherwise double_builder should be checked)

Reply:
this checks were simplified